### PR TITLE
Can no longer immediately return a wayfinding pinpointer for a costume

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request

You can buy a wayfinding pinpointer and immediately return it for a chance at a non-shitty costume spawning. This allows people to circumvent theatre access and costume crates.

This makes it so returning your own pinpointer gives you nothing if it's roundstart, the refund amount is 0 or the dispenser can't afford the refund. Otherwise, you'll be given 2/3 of what you bought it for.

If you return someone else's pinpointer, you'll either be given 2/3 of what they paid or a costume if it's from the quirk, 2/3 is 0 credits or it can't afford to pay out a 2/3 refund.

## Why It's Good For The Game

Less costumes littering arrivals. Less people out of uniform whose jobs you can't tell at a glance. More fighting the clown and/or the mime to use the theatre autodrobe. More costume crate-driven economic interaction. More waiting in line to ask the HoP for theatre access (gripping roleplay)

## Changelog
:cl: cacogen
del: You can't return your own wayfinding pinpointer to get a costume now
code: Improves code a bit in wayfinding.dm
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
> You can buy a wayfinding pinpointer and immediately return it for a chance at a non-shitty costume spawning. This allows people to circumvent theatre access and costume crates.

I thought that non-roundstart pinpointers already only could be refunded for credits (and not costumes) anyway?

**Comment by ATH1909** [comment]
image

**Comment by ATH1909** [comment]
wait a minute, oh, I see the problem here

the if(synth_acc.has_money(refund_amt) && !WP.roundstart) if block is a part of the WP.owner != user.real_name block for some reason. hold on, I can fix that.

**Comment by cacogen** [comment]
No, your own pinpointer can only be refunded for a costume. Which is intentional. I don't remember why. At the moment though the cost is 0 while the refund is more than that.

**Comment by cacogen** [comment]
And in actual fact if you're against this fix because you only care about this object to play dress-up then the fix you did and that I subsequently implemented here is less desirable because now you can only get shitassed credits on bought pinpointers instead of just having to return a couple of minutes later

**Comment by Mothblocks** [comment]
why on earth does it give costumes

**Comment by cacogen** [comment]
It gives them as a reward instead of credits in certain cases to prevent the machine being exploited for credits. I chose costumes because they're fun and harmless. Of course, people have to game the system. 

This was supposed to be a simple fix for a minor issue.

**Comment by Mothblocks** [comment]
> It gives them as a reward instead of credits in certain cases to prevent the machine being exploited for credits

fair enough

File: code/game/objects/items/wayfinding.dm [hunk 1]
```diff
@@ -155,57 +157,48 @@
...
 			var/costume = pick(subtypesof(/obj/effect/spawner/bundle/costume))
 			new costume(user.loc)
-
-		set_expression("veryhappy", 2 SECONDS)
-
-		var/is_a_thing = "are [refund_amt] credit\s."
-		if(!refundiscredits)
 			is_a_thing = "is a freshly synthesised costume!"
 			if(prob(funnyprob))
 				is_a_thing = "is a pulse rifle! Just kidding it's a costume."
```
Comment (Mothblocks) [inline code/game/objects/items/wayfinding.dm:185]: This text is now outdated.

**Comment by Mothblocks** [comment]
>  fix: You can't return your own wayfinding pinpointer to get a costume now

...Is this a fix, or are you trying to cheat GBP?

File: code/game/objects/items/wayfinding.dm [hunk 2]
```diff
@@ -155,57 +157,48 @@
 		user.put_in_hands(P)
 		P.owner = user.real_name
 
-/obj/machinery/pinpointer_dispenser/attackby(obj/item/I, mob/user, params)
+/obj/machinery/pinpointer_dispenser/attackby(obj/item/I, mob/living/user, params)
 	if(machine_stat & (BROKEN|NOPOWER))
 		return ..()
 
 	if(istype(I, /obj/item/pinpointer/wayfinding))
 		var/obj/item/pinpointer/wayfinding/WP = I
```
Comment (Mothblocks) [inline code/game/objects/items/wayfinding.dm:165]: You've already basically touched all the code that used this--take this time to give it a better variable name.

File: code/game/objects/items/wayfinding.dm [hunk 3]
```diff
@@ -155,57 +157,48 @@
...
 			recycling = "feeding me"
 
-		//To imply they got a costume instead of money because it was their pinpointer they recycled
-		var/the_pinpointer = "your pinpointer"
+		var/the_pinpointer = "your pinpointer" //To imply they got a costume because it was their pinpointer
 		if(!itsmypinpointer)
 			the_pinpointer = "that pinpointer"
 
-		say("<span class='robot'>Thank you for [recycling] [the_pinpointer]! Here [is_a_thing]</span>")
+		to_chat(user, "<span class='notice'>You put \the [WP] in the return slot.</span>")
```
Comment (Mothblocks) [inline code/game/objects/items/wayfinding.dm:197]:
```suggestion
		to_chat(user, "<span class='notice'>You put [WP] in the return slot.</span>")
```

File: code/game/objects/items/wayfinding.dm [hunk 4]
```diff
@@ -155,57 +157,48 @@
...
-			qdel(WP)
+		var/is_a_thing = "are [refund_amt] credit\s."
+		if(refund_amt > 0 && synth_acc.has_money(refund_amt) && !WP.roundstart)
+			synth_acc._adjust_money(-refund_amt)
+			var/obj/item/holochip/holochip = new (user.loc)
+			holochip.credits = refund_amt
+			holochip.name = "[holochip.credits] credit holochip"
+			if(ishuman(user))
+				var/mob/living/carbon/human/customer = user
+				customer.put_in_hands(holochip)
```
Comment (Mothblocks) [inline code/game/objects/items/wayfinding.dm:179]: `put_in_hands` is a proc on `/mob`, what's with the typecast?

File: code/game/objects/items/wayfinding.dm [hunk 5]
```diff
@@ -155,57 +157,48 @@
...
-				holochip.name = "[holochip.credits] credit holochip"
-				if(ishuman(user))
-					var/mob/living/carbon/human/customer = user
-					customer.put_in_hands(holochip)
-
-		if(!refundiscredits)
-			qdel(WP)
+		var/is_a_thing = "are [refund_amt] credit\s."
+		if(refund_amt > 0 && synth_acc.has_money(refund_amt) && !WP.roundstart)
+			synth_acc._adjust_money(-refund_amt)
```
Comment (Mothblocks) [inline code/game/objects/items/wayfinding.dm:173]: Don't call internal methods like this.

```suggestion
			synth_acc.adjust_money(-refund_amt)
```

**Comment by cacogen** [comment]
>...Is this a fix, or are you trying to cheat GBP?

I see this as an exploit. The reason I made this was because I saw it being taken advantage of on the live server and it's something I worked hard on. So no, me labelling it fix wasn't insincere. I can change it if you want though.

File: code/game/objects/items/wayfinding.dm (hunk 5 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:173]: Isn't that what the proc is called?

File: code/game/objects/items/wayfinding.dm (hunk 4 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:179]: Was it always like that, though? I first coded this a year ago.

File: code/game/objects/items/wayfinding.dm (hunk 5 above)
Comment (Mothblocks) [reply code/game/objects/items/wayfinding.dm:173]: No.

File: code/game/objects/items/wayfinding.dm (hunk 4 above)
Comment (Mothblocks) [reply code/game/objects/items/wayfinding.dm:179]: Yes.

File: code/game/objects/items/wayfinding.dm (hunk 2 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:165]: ok

File: code/game/objects/items/wayfinding.dm (hunk 5 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:173]: You seem to have a lot of assumptions driving your reaction to this PR and how you're responding to me at the moment.

That's Goof's proc he added when he made economy a couple of years ago.

File: code/game/objects/items/wayfinding.dm (hunk 4 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:179]: I'm not sure if that's true, but I'm not at the computer to check at the moment.

**Comment by cacogen** [comment]
What a petty thing to do.

File: code/game/objects/items/wayfinding.dm (hunk 1 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:185]: I don't understand what you mean by that.

File: code/game/objects/items/wayfinding.dm (hunk 4 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:179]: So I can't find any evidence this changed looking at the proc's blame but I'm not sure why I'd type it unless it had runtimed first. I remember reading something about a change made for mobs without hands or something. Anyway, I'll remove it.

Comment (Mothblocks) [reply code/game/objects/items/wayfinding.dm:179]: image

File: code/game/objects/items/wayfinding.dm (hunk 5 above)
Comment (Mothblocks) [reply code/game/objects/items/wayfinding.dm:173]: I don't know what you're talking about. `_adjust_money` is the internal proc--`adjust_money` is the one you should be calling.

image

File: code/game/objects/items/wayfinding.dm (hunk 1 above)
Comment (Mothblocks) [reply code/game/objects/items/wayfinding.dm:183]: Nevermind, my mistake--didn't see costumes were still occasionally being dispensed.

File: code/game/objects/items/wayfinding.dm (hunk 4 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm]: Yes, I already said that.

File: code/game/objects/items/wayfinding.dm (hunk 5 above)
Comment (cacogen) [reply code/game/objects/items/wayfinding.dm:173]: I was wrong but I still don't think you're coming at this from a fair-minded place. I haven't had any issues with other maintainers. So you trying to generalise it and make it only about me isn't fair.

Comment (Mothblocks) [reply code/game/objects/items/wayfinding.dm]: I have no idea what you're talking about, I didn't even know this feature existed until today, I don't have any biases towards it or you.

Comment (cacogen) [reply code/game/objects/items/wayfinding.dm]: Maybe I'm just reading something into it that isn't there then.

**Comment by Mothblocks** [comment]
Please undraft your PR if it's ready for merge.

**Pull request merged**